### PR TITLE
release the rust-sdk to crates.io

### PR DESCRIPTION
**Pull request by bmwill**
Once this PR lands the following crates will be published to crates.io:
* move-core-types
* diem-crypto-derive
* diem-crypto
* diem-types
* diem-transaction-builder
* diem-json-rpc-types
* diem-client
* diem-sdk

All of these crates will initially be versioned by `0.0.X` where each subsequent release will bump `X`. This is due to the fact that while we intend to have the core functionality be compatible with any current running networks (e.g. testnet) we aren't prepared to commit to a more stable API exposed from these crates. As such each release will be a breaking change semver release until we're able to better clean up and stabilize the APIs exported from these packages.

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-transaction-builder 0.0.1 (workspace, path 'client/transaction-builder')
    * features: default
  M diem-client 0.0.1 (workspace, path 'sdk/client')
    * version DOWNGRADED from 0.1.0
    * added features: ipnet
    * removed features: proxy
    * (unchanged features: blocking, ureq)
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-json-rpc-types 0.0.1 (workspace, path 'json-rpc/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  R transaction-builder-generated 0.1.0 (workspace, path 'client/transaction-builder')
    * (old features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-transaction-builder 0.0.1 (workspace, path 'client/transaction-builder')
    * features: default
  M diem-client 0.0.1 (workspace, path 'sdk/client')
    * version DOWNGRADED from 0.1.0
    * added features: ipnet
    * removed features: proxy
    * (unchanged features: blocking, ureq)
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-json-rpc-types 0.0.1 (workspace, path 'json-rpc/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  R transaction-builder-generated 0.1.0 (workspace, path 'client/transaction-builder')
    * (old features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

File: sdk/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "diem-sdk"
+version = "0.0.0"
```
Comment (xli) [inline sdk/Cargo.toml:3]: 0.0.1?

**Review by xli — APPROVED** [review]

File: sdk/Cargo.toml (hunk 1 above)
Comment (bmwill) [reply sdk/Cargo.toml:3]: Thanks for catching that :)

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-transaction-builder 0.0.1 (workspace, path 'client/transaction-builder')
    * features: default
  M diem-client 0.0.1 (workspace, path 'sdk/client')
    * version DOWNGRADED from 0.1.0
    * added features: ipnet
    * removed features: proxy
    * (unchanged features: blocking, ureq)
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-json-rpc-types 0.0.1 (workspace, path 'json-rpc/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  R transaction-builder-generated 0.1.0 (workspace, path 'client/transaction-builder')
    * (old features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-transaction-builder 0.0.1 (workspace, path 'client/transaction-builder')
    * features: default
  M diem-client 0.0.1 (workspace, path 'sdk/client')
    * version DOWNGRADED from 0.1.0
    * added features: ipnet
    * removed features: proxy
    * (unchanged features: blocking, ureq)
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-json-rpc-types 0.0.1 (workspace, path 'json-rpc/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  R transaction-builder-generated 0.1.0 (workspace, path 'client/transaction-builder')
    * (old features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

File: devtools/x/src/lint/workspace_classify.rs [hunk 2]
```diff
@@ -121,13 +121,16 @@ impl<'cfg> PackageLinter for DefaultOrTestOnly<'cfg> {
...
-                    "library package, listed in default-members:
-                     * if test-only, add to test-only in x.toml instead
-                     * otherwise, remove it from default-members and make it a dependency of a binary"
-                );
-                out.write(LintLevel::Error, msg);
+                if matches!(package.publish(), Some(&[])) {
+                    // Library, listed in default members. It shouldn't be.
+                    // unless it is a published library
+                    let msg = indoc!(
+                        "library package, listed in default-members:
```
Comment (metajack) [inline devtools/x/src/lint/workspace_classify.rs:128]: Should the message mention that it's unpublished?

File: sdk/client/src/blocking.rs [hunk 3]
```diff
@@ -23,6 +21,10 @@ use diem_types::{
     transaction::{SignedTransaction, Transaction},
 };
 use serde::{de::DeserializeOwned, Serialize};
+use std::time::Duration;
+
+#[path = "../../../common/proxy/src/lib.rs"]
```
Comment (metajack) [inline sdk/client/src/blocking.rs:26]: I think you should comment about this here, since it is a little bizarre.

**Review by metajack — APPROVED** [review]
I get that you need to move workspace-hack to dev-deps in order to have crates.io ignore it, but won't this mean it won't be as effective? It will no longer normalize 3p dep feature edges on normal dependencies right?

Is there some other way we plan to mitigate this?

Looks great other than that.

**Comment by bmwill** [comment]
> I get that you need to move workspace-hack to dev-deps in order to have crates.io ignore it, but won't this mean it won't be
> as effective? It will no longer normalize 3p dep feature edges on normal dependencies right?
> Is there some other way we plan to mitigate this?

Yes that's true that it won't be as effective if you're not running test builds, but given that its a currently small subset of crates I figured it would be an acceptable loss. Really the solution would be to no include workspace-hack in any tomls and instead have x implicitly add it during builds.

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-transaction-builder 0.0.1 (workspace, path 'client/transaction-builder')
    * features: default
  M diem-client 0.0.1 (workspace, path 'sdk/client')
    * version DOWNGRADED from 0.1.0
    * added features: ipnet
    * removed features: proxy
    * (unchanged features: blocking, ureq)
  M diem-crypto 0.0.1 (workspace, path 'crypto/crypto')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default, fiat)
  M diem-json-rpc-types 0.0.1 (workspace, path 'json-rpc/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
  M diem-types 0.0.1 (workspace, path 'types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  M move-core-types 0.0.1 (workspace, path 'language/move-core/types')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: default)
  R transaction-builder-generated 0.1.0 (workspace, path 'client/transaction-builder')
    * (old features: default)

host packages:
  M diem-crypto-derive 0.0.1 (workspace, path 'crypto/crypto-derive')
    * version DOWNGRADED from 0.1.0
    * (unchanged features: [none])
```

**Pull request merged**